### ISSUE TITLE
Navigation not working for Android 

### ISSUE THREAD

**Issue by tomnorg** [xamarinhq/xamu-infrastructure#18]
This issue is related / extension to question asked by @crazynitha [`https://github.com/xamarinhq/xamu-infrastructure/issues/17`](here)

I am taking the same example provided by Nitha for illustrating the issue. 

Steps to reproduce. 
1. launch the application.
2. Click on Navigate to detail page button. 
3. Navigation works correctly.
4. Come back to Main page.
5. Click the **back button** and exit the app to  background.
6. Re-launch the app by clicking the app icon.
7. Launches the main screen.
8. Click on Navigate to detail page button again.
9. Navigation not working.

I have implemented both the method's you mentioned in the [https://github.com/xamarinhq/xamu-infrastructure/issues/17](issue17). Copying the same for your reference. But both are not working from me.

**First Method with static Constructor**

```
using Xamarin.Forms;
using Xamarin.Forms.Xaml;
using XamarinUniInfraApp.Core;
using XamarinUniInfraApp.Core.ViewModels;
using XamarinUniversity.Infrastructure;
using XamarinUniversity.Services;

[assembly: XamlCompilation(XamlCompilationOptions.Compile)]
namespace XamarinUniInfraApp
{
    public partial class App : Application
    {
        public App()
        {
            InitializeComponent();
            MainPage = new NavigationPage(new MainPage());
        }
        static App()
        {
            var dependencyService = XamUInfrastructure.Init();

            // Register View Models
            dependencyService.Register<MainViewModel>();
            dependencyService.Register<DetailViewModel>();

            // Register pages with nav service
            RegisterPages(dependencyService.Get<INavigationPageService>());
        }

        private static void RegisterPages(INavigationPageService navService)
        {
            navService?.RegisterPage(AppPages.Main, () => new MainPage());
            navService?.RegisterPage(AppPages.Detail, () => new DetailPage());
        }

        protected override void OnStart()
        {
            // Handle when your app starts
        }

        protected override void OnSleep()
        {
            // Handle when your app sleeps
        }

        protected override void OnResume()
        {
            // Handle when your app resumes
        }
    }
}
```

**Method with static bool variable**

```
using Xamarin.Forms;
using Xamarin.Forms.Xaml;
using XamarinUniInfraApp.Core;
using XamarinUniInfraApp.Core.ViewModels;
using XamarinUniversity.Infrastructure;
using XamarinUniversity.Services;

[assembly: XamlCompilation(XamlCompilationOptions.Compile)]
namespace XamarinUniInfraApp
{
    public partial class App : Application
    {
        private static bool _isInitialized;
        public App()
        {
            if (!_isInitialized)
            {
                var dependencyService = XamUInfrastructure.Init();

                // Register View Models
                dependencyService.Register<MainViewModel>();
                dependencyService.Register<DetailViewModel>();

                // Register pages with nav service
                RegisterPages(dependencyService.Get<INavigationPageService>());
                _isInitialized = true;
            }
            InitializeComponent();
            MainPage = new NavigationPage(new MainPage());
        }

        private void RegisterPages(INavigationPageService navService)
        {
            navService?.RegisterPage(AppPages.Main, () => new MainPage());
            navService?.RegisterPage(AppPages.Detail, () => new DetailPage());
        }

        protected override void OnStart()
        {
            // Handle when your app starts
        }

        protected override void OnSleep()
        {
            // Handle when your app sleeps
        }

        protected override void OnResume()
        {
            // Handle when your app resumes
        }
    }
}
```

[Link to Sample App](https://1drv.ms/u/s!AucxlUaT1vDkacn4S3O2GzHsQes)

**Comment by tomnorg** [comment]
@markjulmar & @patridge : Requesting your help.

**Comment by tomnorg** [comment]
@markjulmar, Need your help. I am waiting for an answer since two weeks. Please help.

**Comment by markjulmar** [comment]
@tomnorg  -- I tried this example in a new app and had no trouble with it. I'll pull down your specific sample and look through it. Here's what I did:

1. Launch app
2. Navigate to second page (detail)
3. Go back.
4. Home button
5. Launch app
6. Navigate to second page (detail).

It all works. Are you rotating the device or anything?

One note: 

There's no need to register ViewModels. The DependencyService will create the type directly if it's not registered. The only reason to register types is if you have an abstraction for it. So you should be able to do this:

```csharp
static App()
{
    var dependencyService = XamUInfrastructure.Init();
    RegisterPages(dependencyService.Get<INavigationPageService>());
}
```

and it should all still work fine.

**Comment by markjulmar** [comment]
Oh, and the static ctor is the way to go for this.

**Comment by tomnorg** [comment]
Thanks for you response,

I am not rotating the device, It is straight forward flow. And I was able to reproduce it. 

Launch app and go to second page. Then come back to  first page again press back press and exit the app. From that point If i try to launch app again and hit the navigation button to second page. Is is not navigating for me. :(

**There's no need to register ViewModels.** - If I remove VM registration code, the VM instantiation is not happening.   

 **I tried this example in a new app** - Could you please share the app that you have tried? So that I can verify the same in my device and if required I can share a video for you as well. 

As I mentioned in the main thread, I was using the sample app created by  @crazynitha . So I am not sure everything is configured in correct way or not.

Need your help to figure out the issue. 

regads,
Tom

**Comment by markjulmar** [comment]
I see the issue. Since the `NavigationPage` is being recreated when you go through that flow, it messes up the navigation service. I think we can fix this in the library code - let me give it a shot.

**Comment by markjulmar** [comment]
Ok, this is fixed in 2.2 of the library. I just published it and verified it with your app. I did make a few changes to validate it.

First, I added a few properties to the `DetailsViewModel` - one is a `Text` property which displays the changing (underlying) navigation property. That was _not_ changing before which was the issue you hit.

Second, I added an alternative approach to using the DependencyService extension which might be more to your liking to the MainPage + MainViewModel. Here are the basics:

1. Have the `MainPage` constructor take the VM as a parameter.
2. Change the `RegisterPage` calls to go through the dependency service so they get inflated with parameters automatically.

```csharp
public partial class MainPage : ContentPage
{
   public MainPage(MainViewModel viewModel)
   {
      BindingContext = viewModel;
      InitializeComponent();
   }
}
```

```csharp
navService.RegisterPage(AppPages.Main, dependencyService.Get<MainPage>);
```

I used the dependency service to "get" the main page - it's not registered at all - but because we are asking for a concrete class, the wrapper will use reflection to create it and _pass_ in the view model automatically. In fact, you can have parameters be passed to the ViewModel too, both of your ViewModels take the dependency service:

```csharp
public MainViewModel(IDependencyService serviceLocator)
{
}
```

The `DetailViewModel` is using the updated `DependencyServiceExtension` which now goes through our wrapper (vs. the underlying Forms version) so it also can inflate objects with parameters.

Here's the [updated sample](https://www.dropbox.com/s/7p0o9d35vqdt2ev/XamUApp2-test.zip?dl=0).

**Comment by tomnorg** [comment]
Thanks for the detailed explanation & quick fix.
